src/sage/categories/rings.py

```diff
@@ -1910,6 +1912,37 @@ def _divide_if_possible(self, y):
                 raise ValueError("%s is not divisible by %s" % (self, y))
             return q
 
+    class CartesianProducts(CartesianProductsCategory):
+        class ParentMethods:
+            @property
+            @cached_method
+            def __nonzero_factors(self):
+                return tuple(s for s in self._sets if not s.is_zero())
```

Comment:
```suggestion
            def __nonzero_factors(self):
                return tuple([s for s in self._sets if not s.is_zero()])
```
This also needs a docstring/test.

Reply:
does pre-creating a list really help with anything (you need to iterate over it anyway, and if you pre-create a list then you need to copy it from the list to the tuple afterwards)?

Reply:
It's marginally faster (surprising), so I try to use it as a general practice.

Reply:
surprising indeed.


Python 3.10

```
In [1]: %timeit tuple(x**2 for x in range(10))
2.41 µs ± 15.7 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [2]: %timeit tuple([x**2 for x in range(10)])
2.34 µs ± 14.2 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [3]: %timeit tuple(x**2 for x in range(1000))
289 µs ± 1.1 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [4]: %timeit tuple([x**2 for x in range(1000)])
258 µs ± 27.1 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

Python 3.12 or 3.13

```
In [1]: %timeit tuple(x**2 for x in range(10))
540 ns ± 52 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)

In [2]: %timeit tuple([x**2 for x in range(10)])
277 ns ± 0.345 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)

In [3]: %timeit tuple(x**2 for x in range(1000))
44.1 μs ± 5.06 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

In [4]: %timeit tuple([x**2 for x in range(1000)])
34 μs ± 44.3 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```

Though I see nothing prevent Python from implementing `tuple(iterator)` to be as fast as `tuple([*iterator])`, so a better course of action would be to propose the speedup to Python.

After the speedup, I'd expect the `tuple(iterator)` to be marginally faster than `tuple([*iterator])`, so such optimization attempts would "backfire" in some way.

Reply:
actually it's probably because `tuple` need to lookup the global namespace.

```
In [6]: %timeit (*(x**2 for x in range(1000)),)
241 µs ± 10.6 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

even faster.

strangely enough it doesn't explain the difference in Python 3.12

```
In [6]: %timeit (*(x**2 for x in range(1000)),)
47.2 μs ± 1.18 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

In [7]: %timeit (*[x**2 for x in range(1000)],)
36.5 μs ± 77.8 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

In [8]: %timeit tuple([x**2 for x in range(1000)])
33.8 μs ± 710 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```

Reply:
anyway, resolved, there's virtually no difference in readability/maintainability anyway. I open https://github.com/python/cpython/issues/138325 for the performance issue.